### PR TITLE
chore: remove unused rust dependencies

### PR DESCRIPTION
**Pull request by ForsakenHarmony**
### What?

Toolchain is updated as well.

Should improve compile times marginally, also added the new parallel frontend.

Depends on https://github.com/vercel/turbo/pull/7409

Closes PACK-2526

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 265019ffec26bda19c3243b011041ef95949049d

`pnpm test-start test/e2e/app-dir/navigation/navigation.test.ts`
- app dir - navigation > navigation between pages and app > should not continously initiate a mpa navigation to the same URL when router state changes

<details>
<summary>Expand output</summary>

  ● app dir - navigation › navigation between pages and app › should not continously initiate a mpa navigation to the same URL when router state changes

    page.waitForSelector: Timeout 60000ms exceeded.
    Call log:
      - waiting for locator('#link-to-slow-page')
      -   locator resolved to visible <a href="/slow-page" id="link-to-slow-page">To /slow-page</a>

      421 |     return this.chain(() => {
      422 |       return page
    > 423 |         .waitForSelector(selector, { timeout, state: 'attached' })
          |          ^
      424 |         .then(async (el) => {
      425 |           // it seems selenium waits longer and tests rely on this behavior
      426 |           // so we wait for the load event fire before returning

      at waitForSelector (lib/browsers/playwright.ts:423:10)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/unused-dependencies | Change |
| - | - | - | - |
| buildDuration | 20s | 20.3s | ⚠️ +245ms |
| buildDurationCached | 7.8s | 7.1s | N/A |
| nodeModulesSize | 196 MB | 196 MB | ✓ |
| nextStartRea..uration (ms) | 427ms | 435ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/unused-dependencies | Change |
| - | - | - | - |
| 1068-HASH.js gzip | 30.3 kB | 30.3 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.5 kB | 53.5 kB | N/A |
| 4944-HASH.js gzip | 5.04 kB | 5.03 kB | N/A |
| 8423.HASH.js gzip | 181 B | 181 B | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 241 B | 244 B | N/A |
| main-HASH.js gzip | 32.1 kB | 32.1 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 47.1 kB | 47.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/unused-dependencies | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/unused-dependencies | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 196 B | ✓ |
| _error-HASH.js gzip | 184 B | 183 B | N/A |
| amp-HASH.js gzip | 503 B | 504 B | N/A |
| css-HASH.js gzip | 323 B | 324 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.51 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 259 B | N/A |
| head-HASH.js gzip | 353 B | 351 B | N/A |
| hooks-HASH.js gzip | 370 B | 370 B | ✓ |
| image-HASH.js gzip | 4.21 kB | 4.2 kB | N/A |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.68 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 313 B | 314 B | N/A |
| script-HASH.js gzip | 386 B | 385 B | N/A |
| withRouter-HASH.js gzip | 309 B | 311 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 931 B | 931 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/unused-dependencies | Change |
| - | - | - | - |
| _buildManifest.js gzip | 485 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/unused-dependencies | Change |
| - | - | - | - |
| index.html gzip | 528 B | 528 B | ✓ |
| link.html gzip | 540 B | 541 B | N/A |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 528 B | 528 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/unused-dependencies | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94.5 kB | 94.5 kB | N/A |
| page.js gzip | 151 kB | 151 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/unused-dependencies | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 626 B | 625 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 44.6 kB | 44.6 kB | N/A |
| edge-runtime..pack.js gzip | 1.94 kB | 1.94 kB | ✓ |
| Overall change | 2.1 kB | 2.1 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/unused-dependencies | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 166 kB | 166 kB | ✓ |
| app-page-exp..prod.js gzip | 95.9 kB | 95.9 kB | ✓ |
| app-page-tur..prod.js gzip | 97.6 kB | 97.6 kB | ✓ |
| app-page-tur..prod.js gzip | 92 kB | 92 kB | ✓ |
| app-page.run...dev.js gzip | 136 kB | 136 kB | ✓ |
| app-page.run..prod.js gzip | 90.6 kB | 90.6 kB | ✓ |
| app-route-ex...dev.js gzip | 22 kB | 22 kB | ✓ |
| app-route-ex..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.6 kB | 14.6 kB | ✓ |
| app-route.ru...dev.js gzip | 21.7 kB | 21.7 kB | ✓ |
| app-route.ru..prod.js gzip | 14.6 kB | 14.6 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.47 kB | 9.47 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.74 kB | 9.74 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.47 kB | 9.47 kB | ✓ |
| pages-turbo...prod.js gzip | 22.1 kB | 22.1 kB | ✓ |
| pages.runtim...dev.js gzip | 22.7 kB | 22.7 kB | ✓ |
| pages.runtim..prod.js gzip | 22.1 kB | 22.1 kB | ✓ |
| server.runti..prod.js gzip | 50.1 kB | 50.1 kB | ✓ |
| Overall change | 927 kB | 927 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/unused-dependencies | Change |
| - | - | - | - |
| 0.pack gzip | 1.55 MB | 1.55 MB | N/A |
| index.pack gzip | 103 kB | 103 kB | ⚠️ +374 B |
| Overall change | 103 kB | 103 kB | ⚠️ +374 B |

</details>

</details>
Commit: 265019ffec26bda19c3243b011041ef95949049d

File: .cargo/config.toml [hunk 1]
```diff
@@ -2,15 +2,25 @@
 CARGO_WORKSPACE_DIR = { value = "", relative = true }
 
 [build]
-
+rustflags = [
+  "--cfg",
+  "tokio_unstable",
+  "-Zshare-generics=y", # make the current crate share its generic instantiations
+  "-Zthreads=8", # parallel frontend https://blog.rust-lang.org/2023/11/09/parallel-rustc.html
```
Comment (bgw) [inline .cargo/config.toml:9]: Nice!

Given that this is still pretty experimental and has known bugs, I think it'd be better to do this in a separate commit/PR so that if we run into issues and have to bisect, it's easier to isolate.

The blog post says:

> In multi-threaded mode there are some known bugs, including deadlocks. If compilation hangs, you have probably hit one of them.

File: rust-toolchain.toml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+[toolchain]
+channel = "nightly-2024-04-03"
+components = ["rustfmt", "clippy"]
```
Comment (bgw) [inline rust-toolchain.toml:3]: If we're touching this anyways, should we also add `rust-analyzer`? It shouldn't cause any problems, just increased disk usage for people who might not want it. It's a bit annoying for me on every toolchain update to have to manually enable this component.

**Review by bgw — APPROVED** [review]

**Comment by sokra** [comment]
Native build: https://github.com/vercel/next.js/actions/runs/8691793652

**Review by sokra — DISMISSED** [review]
Native build fails

File: rust-toolchain.toml (hunk 2 above)
Comment (arlyon) [reply rust-toolchain.toml:3]: Are you on vim/emacs? I use zed/vscode which I _believe_ both just bundle it themselves.

**Comment by ForsakenHarmony** [comment]
Native build: https://github.com/vercel/next.js/actions/runs/8695176317
